src/ui/public/courier/data_source/search_source.js

```diff
@@ -51,280 +51,645 @@
+
+function isIndexPattern(val) {
+  return Boolean(val && typeof val.toIndexList === 'function');
+}
+
+export function SearchSourceProvider(Promise, PromiseEmitter, Private, config) {
   const SearchRequest = Private(SearchRequestProvider);
   const SegmentedRequest = Private(SegmentedRequestProvider);
   const normalizeSortRequest = Private(NormalizeSortRequestProvider);
+  const fetchSoon = Private(FetchSoonProvider);
+  const buildESQuery = Private(BuildESQueryProvider);
+  const { fieldWildcardFilter } = Private(FieldWildcardProvider);
+  const getConfig = (...args) => config.get(...args);
 
   const forIp = Symbol('for which index pattern?');
 
-  function isIndexPattern(val) {
-    return Boolean(val && typeof val.toIndexList === 'function');
-  }
+  class SearchSource {
+    constructor(initialState) {
+      this._instanceid = _.uniqueId('data_source');
+
+      this._state = parseInitialState(initialState);
+
+      /**
+       * List of the editable state properties that turn into a
+       * chainable API
+       *
+       * @type {Array}
+       */
+      this._methods = [
+        'type',
+        'query',
+        'filter',
+        'sort',
+        'highlight',
+        'highlightAll',
+        'aggs',
+        'from',
+        'searchAfter',
+        'size',
+        'source',
+        'version',
+        'fields'
+      ];
+
+      // set internal state values
+      this._methods.forEach(name => {
+        this[name] = val => {
+          if (val == null) {
+            delete this._state[name];
+          } else {
+            this._state[name] = val;
+          }
+
+          return this;
+        };
+      });
+
+      this.history = [];
+      this._requestStartHandlers = [];
+
+      this._filterPredicates = [
+        (filter) => {
+          // remove null/undefined filters
+          return filter;
+        },
+        (filter) => {
+          const disabled = _.get(filter, 'meta.disabled');
+          return disabled === undefined || disabled === false;
+        },
+        (filter, state) => {
+          if (!config.get('courier:ignoreFilterIfFieldNotInIndex')) {
+            return true;
+          }
+
+          if ('meta' in filter && 'index' in state) {
+            const field = state.index.fields.byName[filter.meta.key];
+            if (!field) return false;
+          }
+          return true;
+        }
+      ];
+    }
 
-  _.class(SearchSource).inherits(SourceAbstract);
-  function SearchSource(initialState) {
-    SearchSource.Super.call(this, initialState);
-  }
+    /*****
+     * PUBLIC API
+     *****/
+
+
+    index(indexPattern) {
+      const state = this._state;
+
+      const hasSource = state.source;
+      const sourceCameFromIp = hasSource && state.source.hasOwnProperty(forIp);
+      const sourceIsForOurIp = sourceCameFromIp && state.source[forIp] === state.index;
+      if (sourceIsForOurIp) {
+        delete state.source;
+      }
+
+      if (indexPattern === undefined) return state.index;
```

Comment:
This distinction between undefined and null seems brittle to me. One returns the index, and one deletes it, but null and undefined are often treated as synonymous by developers. :/

Reply:
So strange indeed.  I haven't had a chance to really investigate why this is, but don't want to get into it for this PR.  As long as I haven't changed any logic in this PR, my only goal is to get rid of the unnecessary abstraction and turn into es6 style class.